nearswap/src/pool.rs

```diff
@@ -170,3 +169,178 @@ impl Pool {
+    }
+
+    fn expected_added_liquidity(
+        ynear: u128, max_tokens: u128, pool: &Pool
+    ) -> (u128, u128, u128) {
+        let ynear_256 = u256::from(ynear);
+        let p_ynear_256 = u256::from(pool.ynear);
+        let mut added_tokens = (ynear_256 * u256::from(pool.tokens) / p_ynear_256 + 1).as_u128();
+        let shares_minted;
+        let added_near;
+
+        // Adjust near according to max_tokens
+        if max_tokens < added_tokens {
+            added_near = ((u256::from(max_tokens) * p_ynear_256) / u256::from(pool.tokens) + 1)
+                .as_u128();
+            added_tokens = max_tokens;
+            shares_minted = (u256::from(added_near) * u256::from(pool.total_shares)
+                / p_ynear_256)
+                .as_u128();
+        } else {
+            added_near = ynear;
+            shares_minted = (ynear_256 * u256::from(pool.total_shares) / p_ynear_256).as_u128();
+        }
+        return (added_near, added_tokens, shares_minted);
+    }
+
+    fn expected_withdraw(shares: u128, pool: &Pool) -> (u128, u128) {
+        let total_shares2 = u256::from(pool.total_shares);
+        let shares2 = u256::from(shares);
+        let ynear = (shares2 * u256::from(pool.ynear) / total_shares2).as_u128();
+        let token_amount = (shares2 * u256::from(pool.tokens) / total_shares2).as_u128();
+        return(ynear, token_amount);
+    }
+
+    // Empty pool
+    #[test]
+    fn new_pool() {
+        let pool: Pool = setup_pool();
+
+        assert!(pool.ynear == 0, "Pool is not empty");
+        assert!(pool.tokens == 0, "Pool is not empty");
+        assert!(pool.total_shares == 0, "Pool is not empty");
+    }
+
+    // Existing pool
+    #[test]
+    fn add_liquidity_pool() {
+        init_blockchain();
+
+        let caller = "account".to_string();
+        let mut pool: Pool = setup_pool();
+
+        pool.add_liquidity(&caller, 100, 200, 0);
+
+        assert!(pool.ynear == 100, "liquidity added is incorrect");
+        assert!(pool.tokens == 200, "liquidity added is incorrect");
+        assert!(pool.total_shares == 100, "liquidity added is incorrect");
+
+        let (expected_near, expected_tokens, expected_shares) = expected_added_liquidity(200, 400, &pool);
+        
+        // add liquidity again
+        let (near_added, tokens_added, shares_minted) = pool.add_liquidity(&caller, 200, 400, 0);
+
+        assert!(near_added == expected_near, "liquidity added is incorrect");
+        assert!(tokens_added == expected_tokens, "liquidity added is incorrect");
+        assert!(shares_minted == expected_shares, "liquidity added is incorrect");
+
+        let (expected_near2, expected_tokens2, expected_shares2) = expected_added_liquidity(100, 100, &pool);
+        // add liquidity again with ratio 1:1(100:100)
+        let (near_added2, tokens_added2, shares_minted2) = pool.add_liquidity(&caller, 100, 100, 0);
+
+        assert!(near_added2 == expected_near2, "liquidity added is incorrect");
```

Comment:
```suggestion
        assert!(near_added2 == expected_near2, "liquidity added is incorrect");
        assert!(near_added2 == 100, "liquidity added is incorrect");
```

Reply:
shares here would be 51 because pool ratio is (1:2) so user gives 100 tokens therfore near is adjusted to half.

Reply:
right, then let's put 51 here.